### PR TITLE
Move freeing of an old record layer to dtls1_clear_sent_buffer

### PR DESCRIPTION
**Pull request by mattcaswell**
When we are clearing the sent messages queue we should ensure we free any
old write record layers that are no longer in use. Previously this logic
was in dtls1_hm_fragment_free() - but this can end up freeing the current
record layer under certain error conditions.

Fixes https://github.com/openssl/openssl/issues/22664

We also fix an issue during allocation of a new hm_fragment to ensure the data is initialised (commit from @nhorman originall in PR #22677)

### CODE REVIEW

File: ssl/d1_lib.c
Comment (nhorman) [inline ssl/d1_lib.c:143]: if we make this change, does it also not need to be mirrored to dtls1_clear_received_buffer, as that path also relies on dtls1_hm_fragment_free doing this check?

Comment (mattcaswell) [reply ssl/d1_lib.c:143]: No. The "w" in wrlmethod/wrl is for "write". The only place that `msg_header.saved_retransmit_state.wrlmethod` and `msg_header.saved_retransmit_state.wrl` are ever set to anything is in `dtls1_buffer_message` which is explicitly about the `sent_messages` queue. So while `dtls1_hm_fragment_free` gets called from `dtls1_clear_received_buffer` it was never doing anything with the `wrl` because it was always NULL in that code path.

**Comment by mattcaswell** [comment]
#22681 for a very similar patch for 3.1/3.0

**Review by nhorman — APPROVED** [review]

**Review by t-j-h — APPROVED** [review]

File: ssl/d1_lib.c
Comment (mattcaswell) [inline ssl/d1_lib.c:138]: off-by-one alignment

**Comment by mattcaswell** [comment]
Just pushed a trivial fixup to address the alignment issue. I will assume that all approvals still hold.

**Comment by mattcaswell** [comment]
Another fixup pushed as a result of the issue identified in #22681. I think a similar issue is possible here - hence the fixup.

@t8m, @t-j-h - please reconfirm

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
CI failure is unrelated:

````
1 Added function symbol not referenced by debug info:

  [A] EVP_DigestSqueeze@@OPENSSL_3.2.0
````
@slontis

**Comment by mattcaswell** [comment]
> CI failure is unrelated:

Ah this is #22688

**Comment by t8m** [comment]
Ping @t-j-h for reconfirmation after the fixups. Or anybody else please provide the second review.

**Comment by mattcaswell** [comment]
Still needs a second approval or reconfirmation from @t-j-h.

**Comment by mattcaswell** [comment]
Ping for second approval @openssl/committers!

**Review by beldmit — APPROVED** [review]
LGTM

**Pull request closed without merging**
